superset/models/reports.py

```diff
@@ -0,0 +1,172 @@
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=line-too-long,unused-argument,ungrouped-imports
+"""A collection of ORM sqlalchemy models for Superset"""
+import enum
+
+from flask_appbuilder import Model
+from sqlalchemy import (
+    Boolean,
+    Column,
+    DateTime,
+    Float,
+    ForeignKey,
+    Integer,
+    String,
+    Table,
+    Text,
+)
+from sqlalchemy.orm import relationship
+from sqlalchemy.schema import UniqueConstraint
+
+from superset.extensions import security_manager
+from superset.models.core import Database
+from superset.models.dashboard import Dashboard
+from superset.models.helpers import AuditMixinNullable
+from superset.models.slice import Slice
+
+metadata = Model.metadata  # pylint: disable=no-member
+
+
+class ReportScheduleType(str, enum.Enum):
+    ALERT = "Alert"
+    REPORT = "Report"
+
+
+class ReportRecipientType(str, enum.Enum):
+    EMAIL = "Email"
+    SLACK = "Slack"
+
+
+class ReportLogState(str, enum.Enum):
+    SUCCESS = "Success"
+    ERROR = "Error"
+
+
+class ReportEmailFormat(str, enum.Enum):
+    VISUALIZATION = "Visualization"
+    DATA = "Raw data"
+
+
+report_schedule_user = Table(
+    "report_schedule_user",
+    metadata,
+    Column("id", Integer, primary_key=True),
+    Column("user_id", Integer, ForeignKey("ab_user.id"), nullable=False),
+    Column(
+        "report_schedule_id", Integer, ForeignKey("report_schedule.id"), nullable=False
+    ),
+    UniqueConstraint("user_id", "report_schedule_id"),
+)
+
+
+class ReportSchedule(Model, AuditMixinNullable):
+
+    """
+    Report Schedules, supports alerts and reports
+    """
+
+    __tablename__ = "report_schedule"
+    id = Column(Integer, primary_key=True)
+    type = Column(String(50), nullable=False)
+    label = Column(String(150), nullable=False, unique=True)
+    active = Column(Boolean, default=True, index=True)
+    crontab = Column(String(50), nullable=False)
+    sql = Column(Text())
+    # (Reports) M-O to chart
+    chart_id = Column(Integer, ForeignKey("slice.id"), nullable=True)
+    chart = relationship(Slice, backref="report_schedules", foreign_keys=[chart_id])
+    # (Reports) M-O to dashboard
+    dashboard_id = Column(Integer, ForeignKey("dashboards.id"), nullable=True)
+    dashboard = relationship(
+        Dashboard, backref="report_schedules", foreign_keys=[dashboard_id]
+    )
+    # (Alerts) M-O to database
+    database_id = Column(Integer, ForeignKey("dbs.id"), nullable=True)
+    database = relationship(Database, foreign_keys=[dashboard_id])
+    owners = relationship(security_manager.user_model, secondary=report_schedule_user)
+
+    # (Reports) email format
+    email_format = Column(String(50))
```

Comment:
Looks like this may or may not be an email, should this be `content_type`?

Reply:
Good point, I'm tempted on placing it has a key on `recipient_config_json`, better separation of concerns. Yet the out of the box generic, same type for every recipient is a nice to have. What do you think?

Reply:
I agree that this should go in `recipient_config_json` in the related model.